src/vs/workbench/contrib/preferences/browser/settingsTree.ts

```diff
@@ -75,6 +76,88 @@ function getExcludeDisplayValue(element: SettingsTreeSettingElement): IListDataI
 		});
 }
 
+function getEnumOptionsFromSchema(schema: IJSONSchema): IObjectEnumOption[] {
+	const enumDescriptions = schema.enumDescriptions ?? [];
+
+	return (schema.enum ?? []).map((value, idx) => {
+		const description = idx < enumDescriptions.length
+			? enumDescriptions[idx]
+			: undefined;
+
+		return { value, description };
+	});
+}
+
+function getObjectDisplayValue(element: SettingsTreeSettingElement): IObjectDataItem[] {
+	const data = element.isConfigured ?
+		{ ...element.defaultValue, ...element.scopeValue } :
+		element.defaultValue;
+
+	let items: IObjectDataItem[] = [];
+
+	if (element.setting.objectProperties) {
+		const wellDefinedKeys = Object.keys(element.setting.objectProperties);
+		const keyOptions = wellDefinedKeys.map(value => ({ value }));
+
+		items = items.concat(
+			wellDefinedKeys
+				.filter(key => !!data[key])
+				.map(key => {
+					const valueEnumOptions = getEnumOptionsFromSchema(element.setting.objectProperties![key]);
+
+					return {
+						key: {
+							type: 'enum',
+							data: key,
+							options: keyOptions,
+						},
+						value: {
+							type: valueEnumOptions.length > 0 ? 'enum' : 'string',
+							data: data[key],
+							options: valueEnumOptions,
+						},
+					};
+				})
+		);
+	}
+
+	if (element.setting.objectPatternProperties) {
+		const patternsAndSchemas = Object
+			.entries(element.setting.objectPatternProperties)
+			.map(([pattern, schema]) => ({
+				pattern: new RegExp(pattern),
+				schema
+			}));
+
+		const keysWithSchema = Object.keys(data)
+			.filter(key => !!data[key] && !(key in (element.setting.objectProperties ?? {})))
+			.map(key => {
+				const patternAndSchema = patternsAndSchemas.find(({ pattern }) => pattern.test(key));
+				return patternAndSchema
+					? { key, schema: patternAndSchema.schema }
+					: undefined;
+			})
+			.filter(isDefined);
+
+		items = items.concat(keysWithSchema.map(({ key, schema }) => {
+			const valueEnumOptions = getEnumOptionsFromSchema(schema);
+
+			return {
+				key: { type: 'string', data: key },
+				value: {
+					type: valueEnumOptions.length > 0 ? 'enum' : 'string',
+					data: data[key],
+					options: valueEnumOptions,
+				}
+			};
+		}));
+	}
+
+	// TODO @9at8: What should we do if properties don't match?
```

Comment:
We do have validation for settings, that's another thing to think about :)

Reply:
yeah, that's what I was looking into yesterday 😅